### PR TITLE
CA-96: fix: Fix lock related bug

### PR DESCRIPTION
**Pull request by SimonDerejeW**
# PR Review: feat/lock-edge-case → main

**Reviewed Date:** 2026-04-10
**Reviewer:** Claude Code
**PR Branch:** feat/lock-edge-case
**Base Branch:** main

---

## 1. PR SUMMARY

This PR addresses an edge case in the lock functionality for cost estimations. The main changes involve:

1. **Refactoring `EstimationActionsSheet` from StatefulWidget to StatelessWidget** - Moving lock state management from internal widget state to a `ValueNotifier` passed from the parent
2. **Adding BLoC state listener** - Implementing a `BlocListener` to sync lock status changes from the BLoC back to the UI via `ValueNotifier`
3. **Proper resource cleanup** - Using `whenComplete()` to dispose of the `ValueNotifier` when the bottom sheet closes
4. **Test updates** - Updating unit and screenshot tests to work with the new `ValueNotifier`-based approach

The core issue being fixed is ensuring the lock toggle UI stays in sync with the actual BLoC state, particularly when lock operations fail or succeed asynchronously.

---

## REVIEW SUMMARY

| Issue Name | Description | Status | Notes |
|------------|-------------|--------|-------|
| Memory Leak: ValueNotifier Not Guaranteed to Dispose | The `ValueNotifier` is created before `showModalBottomSheet` and disposal is handled in `whenComplete()`, but if the bottom sheet fails to show or throws during creation, the notifier will never be disposed, causing a memory leak. | ✅ | |
| Missing Logs Action in Production Code | The `onLogs` parameter is removed from `EstimationActionsSheet` constructor but the UI still renders the logs action button. The diff shows `onLogs: widget.onLogs,` changed to `onTap: onLogs,` but `onLogs` is no longer a parameter, suggesting this code path will fail at runtime. | ✅ | |
| Test Double Pattern Violation | Tests create `ValueNotifier` but don't verify the actual integration with `BlocListener`. The tests should use a real `ChangeLockStatusBloc` (with fake dependencies) to verify the `BlocListener` correctly updates the `ValueNotifier` when state changes occur. | ✅ | |
| Insufficient Test Coverage for Edge Cases | Tests don't cover critical edge cases: (1) What happens when `ChangeLockStatusFailure` occurs? (2) Does the UI revert to the original state? (3) Is there visual feedback for the user? The BlocListener code suggests it should revert, but no tests verify this behavior. | ✅ | |
| Missing Accessibility Test Updates | The PR modifies a key interactive widget (`EstimationActionsSheet`) but doesn't update or add accessibility tests to verify that the lock toggle's semantics remain accessible when using `ValueListenableBuilder`. |  ✅ | |
| Potential Race Condition in State Restoration | If a user rapidly toggles the lock switch before the previous request completes, multiple `ChangeLockStatusRequested` events could be dispatched. The `BlocListener` will update `lockStatusNotifier` for each completion, potentially causing UI flicker or incorrect final state. | ✅ | |

### CODE REVIEW

File: lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart [hunk 1]
```diff
@@ -88,53 +88,75 @@ class _CostEstimationLandingPageState extends State<CostEstimationLandingPage> {
...
+                onLockToggle: (bool isLocked) {
+                  changeLockStatusBloc.add(
+                    ChangeLockStatusRequested(
+                      estimationId: estimation.id,
+                      isLocked: isLocked,
+                      projectId: widget.projectId,
+                    ),
+                  );
+                },
+                onLogs: () {
```
Comment (ripplearcgit) [inline lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart]: 🔴 **Double-dispose risk.** The `catch` block disposes `lockStatusNotifier`, but `_EstimationActionsSheetWrapperState.dispose()` will also dispose it if the sheet was ever mounted. If `showModalBottomSheet` throws after the wrapper widget mounts, both paths fire → crash.

Remove the `try`/`catch` entirely. Let `_EstimationActionsSheetWrapperState.dispose()` be the single disposal site.

```dart
} catch (e) {
  lockStatusNotifier.dispose(); // ← remove this entire block
}
```

File: lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart [hunk 2]
```diff
@@ -88,53 +88,75 @@ class _CostEstimationLandingPageState extends State<CostEstimationLandingPage> {
...
+            value: changeLockStatusBloc,
+            child: BlocListener<ChangeLockStatusBloc, ChangeLockStatusState>(
+              listener: (context, state) {
+                if (state is ChangeLockStatusSuccess) {
+                  lockStatusNotifier.value = state.isLocked;
+                } else if (state is ChangeLockStatusFailure) {
+                  lockStatusNotifier.value = state.originalValue;
+                }
+              },
+              child: _EstimationActionsSheetWrapper(
```
Comment (ripplearcgit) [inline lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart]: 🔴 **TODO format violation.** Missing `[CA-88]` label and space after colon.

```dart
// TODO: [CA-88] description. https://ripplearc.youtrack.cloud/issue/CA-88
```

File: lib/features/estimation/presentation/widgets/estimation_actions_sheet.dart [hunk 3]
```diff
@@ -2,35 +2,30 @@ import 'package:construculator/libraries/extensions/extensions.dart';
...
-class EstimationActionsSheet extends StatefulWidget {
+class EstimationActionsSheet extends StatelessWidget {
   const EstimationActionsSheet({
     super.key,
     required this.estimationName,
+    required this.lockStatusNotifier,
     required this.onRename,
     required this.onFavourite,
     required this.onRemove,
+    required this.onLockToggle,
```
Comment (ripplearcgit) [inline lib/features/estimation/presentation/widgets/estimation_actions_sheet.dart:13]: 🟡 `lockStatusNotifier` and `onLockToggle` are new public fields without `///` Dart doc comments. All public members require docs per project standard.

```dart
/// The notifier that drives the lock toggle state. Updated externally
/// by [BlocListener] when [ChangeLockStatusBloc] emits a new state.
final ValueNotifier<bool> lockStatusNotifier;

/// Called when the user toggles the lock switch.
final ValueChanged<bool> onLockToggle;
```

**Review by ripplearcgit — CHANGES_REQUESTED** [review]
## Review — CA-96: Fix lock-related edge case

Good approach overall — the `ValueNotifier` + `BlocListener` pattern is clean, the test coverage is thorough, and the `FakeSupabaseWrapper` + `Completer` usage for in-flight delay testing is exactly correct. Two blocking issues before merge.

---

### 🔴 Blocking

**#1 — Double-dispose on `lockStatusNotifier`**

The `ValueNotifier` can be disposed twice: once in the `_EstimationActionsSheetWrapperState.dispose()` and again in the `catch` block of `_showEstimationActionsSheet`. If `showModalBottomSheet` throws *after* the sheet widget has mounted (so the wrapper's `initState` has run), both paths fire, causing a `"A ValueNotifier was used after being disposed"` crash.

The fix is to remove the `catch` block entirely and let `_EstimationActionsSheetWrapperState.dispose()` be the single disposal site. The `try`/`catch` adds no value here — any exception from `showModalBottomSheet` will propagate naturally.

```dart
// ✅ Simplified — no try/catch needed
void _showEstimationActionsSheet(...) async {
  final lockStatusNotifier = ValueNotifier<bool>(estimation.lockStatus.isLocked);
  await showModalBottomSheet(
    ...
    builder: (_) => BlocProvider.value(
      value: changeLockStatusBloc,
      child: BlocListener<ChangeLockStatusBloc, ChangeLockStatusState>(
        listener: ...,
        child: _EstimationActionsSheetWrapper(
          lockStatusNotifier: lockStatusNotifier,
          ...
        ),
      ),
    ),
  );
  // disposal handled exclusively by _EstimationActionsSheetWrapperState.dispose()
}
```

**#2 — TODO format violation**

`// TODO:https://ripplearc.youtrack.cloud/issue/CA-88` is missing the `[CA-88]` label and space. Required format:

```dart
// TODO: [CA-88] description. https://ripplearc.youtrack.cloud/issue/CA-88
```

---

### 🟡 Non-blocking

**#3 — `_EstimationActionsSheetWrapper` adds unnecessary complexity**

Once the double-dispose is fixed (see #1), `_EstimationActionsSheetWrapper` exists solely to call `dispose()` on the notifier. This is an extra `StatefulWidget` + `State` class for a single lifecycle hook. Consider `addPostFrameCallback` + `whenComplete` or simply relying on the BlocListener's containing widget if there's a cleaner host. Not blocking, but worth revisiting.

**#4 — Missing `///` doc comments on new public members**

`lockStatusNotifier` and `onLockToggle` on `EstimationActionsSheet` are new public fields without Dart doc comments. All public members require `///` docs per project standard.

File: lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart [hunk 4]
```diff
@@ -461,3 +480,49 @@ class _CostEstimationLandingPageState extends State<CostEstimationLandingPage> {
...
+      estimationName: widget.estimationName,
+      lockStatusNotifier: widget.lockStatusNotifier,
+      onRename: widget.onRename,
+      onFavourite: widget.onFavourite,
+      onRemove: widget.onRemove,
+      onLockToggle: widget.onLockToggle,
+      onLogs: widget.onLogs,
+    );
+  }
+}
```
Comment (ripplearcgit) [inline lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart]: 🔴 `_EstimationActionsSheetWrapper` exists solely to dispose `lockStatusNotifier`. Since `_showEstimationActionsSheet` is already `async`, dispose the notifier after `await showModalBottomSheet(...)` instead and remove this wrapper class entirely:

```dart
await showModalBottomSheet(...);
lockStatusNotifier.dispose(); // single, guaranteed disposal site
```

**Review by ripplearcgit — CHANGES_REQUESTED** [review]
## Re-review — CA-96: Fix lock-related edge case

Prior blockers #1 (double-dispose) and #2 (TODO CA-88 format) are resolved. One new blocker.

---

### 🔴 Blocking

**#3 — `_EstimationActionsSheetWrapper` is unnecessary**

`_EstimationActionsSheetWrapper` exists solely to call `widget.lockStatusNotifier.dispose()` in its `dispose()` override. This is an entire `StatefulWidget` + `State` class pair for a single lifecycle hook.

`showModalBottomSheet` returns a `Future` that completes when the sheet is dismissed. Since `_showEstimationActionsSheet` is already `async`, disposal can happen directly after the `await` — no wrapper needed:

```dart
) async {
  final changeLockStatusBloc = BlocProvider.of<ChangeLockStatusBloc>(context);
  final lockStatusNotifier = ValueNotifier<bool>(estimation.lockStatus.isLocked);

  await showModalBottomSheet(
    context: context,
    ...
    builder: (_) {
      return BlocProvider.value(
        value: changeLockStatusBloc,
        child: BlocListener<ChangeLockStatusBloc, ChangeLockStatusState>(
          listener: (context, state) {
            if (state is ChangeLockStatusSuccess) {
              lockStatusNotifier.value = state.isLocked;
            } else if (state is ChangeLockStatusFailure) {
              lockStatusNotifier.value = state.originalValue;
            }
          },
          child: EstimationActionsSheet(
            lockStatusNotifier: lockStatusNotifier,
            estimationName: estimation.estimateName,
            onRename: () { ... },
            onFavourite: () { ... },
            onRemove: () { ... },
            onLockToggle: (bool isLocked) { ... },
            onLogs: () { ... },
          ),
        ),
      );
    },
  );

  lockStatusNotifier.dispose(); // ← single, guaranteed disposal site
}
```

Remove `_EstimationActionsSheetWrapper` and its `State` class entirely.

**Review by ripplearcgit — COMMENTED** [review]
### 🟡 Non-blocking

**#2 — TODO CA-472 format**

The CA-472 TODO is still missing the colon after `TODO`. Current:

```dart
// TODO [CA-472] Use `CoreQuickSheet.show`...
```

Required format:

```dart
// TODO: [CA-472] Use `CoreQuickSheet.show` to standardize bottom sheets. https://ripplearc.youtrack.cloud/issue/CA-472
```

File: lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart [hunk 5]
```diff
@@ -88,52 +88,71 @@ class _CostEstimationLandingPageState extends State<CostEstimationLandingPage> {
...
     AppColorsExtension colorTheme,
-  ) {
-    // TODO: https://ripplearc.youtrack.cloud/issue/CA-472/CoreUI-Standardize-bottom-sheets-with-CoreQuickSheet-component (Standardize bottom sheets with CoreQuickSheet component)
-    showModalBottomSheet(
+  ) async {
+    final changeLockStatusBloc = BlocProvider.of<ChangeLockStatusBloc>(context);
+
+    // TODO [CA-472] Use `CoreQuickSheet.show` to standardize bottom sheets  https://ripplearc.youtrack.cloud/issue/CA-472/CoreUI-Standardize-bottom-sheets-with-CoreQuickSheet-component (Standardize bottom sheets with CoreQuickSheet component)
+    await showModalBottomSheet(
       context: context,
```
Comment (ripplearcgit) [inline lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart:99]: 🟡 Missing `:` after `TODO`. Required format:

```dart
// TODO: [CA-472] Use `CoreQuickSheet.show` to standardize bottom sheets. https://ripplearc.youtrack.cloud/issue/CA-472
```

File: lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart [hunk 6]
```diff
@@ -88,53 +88,74 @@ class _CostEstimationLandingPageState extends State<CostEstimationLandingPage> {
...
     CostEstimate estimation,
     AppColorsExtension colorTheme,
-  ) {
-    // TODO: https://ripplearc.youtrack.cloud/issue/CA-472/CoreUI-Standardize-bottom-sheets-with-CoreQuickSheet-component (Standardize bottom sheets with CoreQuickSheet component)
-    showModalBottomSheet(
+  ) async {
+    final changeLockStatusBloc = BlocProvider.of<ChangeLockStatusBloc>(context);
+
+    final lockStatusNotifier = ValueNotifier<bool>(
+      estimation.lockStatus.isLocked,
```
Comment (ripplearcgit) [inline lib/features/estimation/presentation/pages/cost_estimation_landing_page.dart:95]: 🟡 Semicolon instead of colon. Required format:

```dart
// TODO: [CA-472] Use `CoreQuickSheet.show` to standardize bottom sheets. https://ripplearc.youtrack.cloud/issue/CA-472
```

**Review by ripplearcgit — APPROVED** [review]
## Re-review — CA-96: Fix lock-related edge case (Round 3)

All prior blockers resolved. One lingering non-blocking nit.

---

### ✅ Resolved

- **#1 — double-dispose / try-catch**: Removed. `lockStatusNotifier.dispose()` is now the single site after `await showModalBottomSheet(...)`.
- **#3 — `_EstimationActionsSheetWrapper`**: Removed entirely. Clean.
- **#4 — missing `///` docs**: All new public members on `EstimationActionsSheet` are fully documented.

---

### 🟡 Non-blocking

**#2 — TODO CA-472: semicolon instead of colon (third round)**

Current:
```dart
// TODO; [CA-472] Use `CoreQuickSheet.show`...
```

Required:
```dart
// TODO: [CA-472] Use `CoreQuickSheet.show` to standardize bottom sheets. https://ripplearc.youtrack.cloud/issue/CA-472
```

A `;` snuck in where `:` belongs — same convention violation as the previous two rounds, different typo. Fix before merge.

**Comment by SimonDerejeW** [comment]
#runcheck

**Comment by SimonDerejeW** [comment]
#runcheck
